x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java

```diff
@@ -151,8 +167,54 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
 
         Set<String> concreteIndices = new HashSet<>(fieldCapsResponse.getIndexResponses().size());
         for (FieldCapabilitiesIndexResponse ir : fieldCapsResponse.getIndexResponses()) {
-            concreteIndices.add(ir.getIndexName());
+            String indexExpression = ir.getIndexName();
+            concreteIndices.add(indexExpression);
+
+            String clusterAlias = parseClusterAlias(indexExpression);
+            EsqlExecutionInfo.Cluster cluster = executionInfo.getCluster(clusterAlias);
+            // populate the EsqlExecutionInfo cluster map with responses from each cluster coming back from field-caps
+            if (cluster == null) {
+                executionInfo.swapCluster(
+                    clusterAlias,
+                    (k, v) -> new EsqlExecutionInfo.Cluster(clusterAlias, indexExpression, executionInfo.isSkipUnavailable(clusterAlias))
+                );
+            } else {
+                String newIndexExpr = cluster.getIndexExpression() + "," + indexExpression;
+                executionInfo.swapCluster(
+                    clusterAlias,
+                    (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setIndexExpression(newIndexExpr).build()
+                );
+            }
+        }
+
+        // check for "remote unavailable" type errors that occurred during field-caps lookup on remote clusters
+        if (fieldCapsResponse.getFailures() != null) {
+            Set<String> clusterAliasesWithErrors = new HashSet<>();
+            for (FieldCapabilitiesFailure failure : fieldCapsResponse.getFailures()) {
+                if (isRemoteUnavailableException(failure.getException())) {
+                    for (String indexExpression : failure.getIndices()) {
+                        if (indexExpression.indexOf(RemoteClusterAware.REMOTE_CLUSTER_INDEX_SEPARATOR) > 0) {
+                            clusterAliasesWithErrors.add(parseClusterAlias(indexExpression));
+                        }
+                    }
+                }
+            }
+
+            for (String clusterAlias : clusterAliasesWithErrors) {
+                if (executionInfo.getCluster(clusterAlias).isSkipUnavailable()) {
+                    executionInfo.swapCluster(
+                        clusterAlias,
+                        (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SKIPPED).build()
+                    );
+                } else {
+                    // MP FIXME this causes a 500 - is that what we want to return? _search returns 400 when skip_unavailable=false.
+                    throw new IllegalStateException(
+                        "Placeholder error for failing an ESQL search. Required remote cluster could not be contacted:" + clusterAlias
+                    );
```

Comment:
Adding this here now to show how it will work, but the code to handle skip_unavailable (including throwing this error here) will be done in a follow on PR, so I will remove this after the initial draft review. But I wanted to include it here so you can see where it fits in with the execution info design.